packages/flutter/lib/src/painting/matrix_utils.dart

```diff
@@ -136,23 +148,211 @@ class MatrixUtils {
+    //         |          ----                   -----  |
+    //         |              ----          -----       |
+    //         |                  ---- -----            |
+    //         |                      v                 |
+    //         +---------------(+wx+hx,+wy+hy)----------+
+    // ```
+    // In this diagram, consider that:
+    // ```
+    // wx would be a positive number
+    // hx would be a negative number
+    // wy and hy would both be positive numbers
+    // ```
+    // As a result, wx pushes out the right wall, hx pushes out the left wall,
+    // and both wy and hy push down the bottom wall of the bounding box. The
+    // wx,hx pair (of opposite signs) worked on opposite walls and the final
+    // opposite corner had an X coordinate between the limits they established.
+    // The wy,hy pair (of the same sign) both worked together to push the
+    // bottom wall down by their sum.
+    //
+    // This relationship allows us to simply start with the point computed by
+    // transforming the upper left corner of the rectangle, and then
+    // conditionally adding wx, wy, hx, and hy to either the left or top
+    // or right or bottom of the bounding box independently depending on sign.
+    // In that case we only need 4 comparisons and 4 additions total to
+    // compute the bounding box, combined with the 8 multiplications and
+    // 4 additions to compute the transformed point and relative vectors
+    // for a total of 8 multiplies, 8 adds, and 4 comparisons.
+    //
+    // An astute observer will note that we do need to do 2 subtractions at
+    // the top of the method to compute the width and height.  Add those to
+    // all of the relative solutions listed above.  The test for perspective
+    // also adds 3 compares to the affine case and up to 3 compares to the
+    // perspective case (depending on which test fails, the rest are omitted).
+    //
+    // The final tally:
+    // basic method          = 60 mul + 48 add + 12 compare
+    // optimized perspective = 12 mul + 22 add + 15 compare + 2 sub
+    // optimized affine      =  8 mul +  8 add +  7 compare + 2 sub
+    //
+    // Since compares are essentially subtractions and subtractions are
+    // the same cost as adds, we end up with:
+    // basic method          = 60 mul + 60 add/sub/compare
+    // optimized perspective = 12 mul + 39 add/sub/compare
+    // optimized affine      =  8 mul + 17 add/sub/compare
+
+    final double wx = storage[0] * w;
+    final double hx = storage[4] * h;
+    final double rx = storage[0] * x + storage[4] * y + storage[12];
+
+    final double wy = storage[1] * w;
+    final double hy = storage[5] * h;
+    final double ry = storage[1] * x + storage[5] * y + storage[13];
+
+    if (storage[3] == 0.0 && storage[7] == 0.0 && storage[15] == 1.0) {
+      double left  = rx;
+      double right = rx;
+      if (wx < 0) { left += wx; } else { right += wx; }
```

Comment:
These and similar have to be split up on multiple lines I believe.

Reply:
Dang it. I saw the warnings in Android Studio, but was going to come back and see if I could find some other format that would make the logic as easy to follow but comply. Unfortunately, this formatting does a great job of pointing out that we are deciding which "wall" to add the vector component to and splitting it up just wastes lines of code to obscure that logic.

`(wx < 0) ? left += wx : right += wx;`
works fine, passes the analyzer, is uglier than what I wrote and slightly misleading in that the value is discarded, but at least maintains making the logic visible.

Grrr...

Reply:
(And it would be interesting if we could write `(wx < 0 ? left : right) += wx;` but ternary operators can't be LHS of an assignment.)

Reply:
Another alternative:
```
      double left  = rx + (wx > 0 ? 0 : wx) + (hx > 0 ? 0 : hx);
      double right = rx + (wx < 0 ? 0 : wx) + (hx < 0 ? 0 : hx);
```
But it leads to more work unless the compiler is smart about how it evaluates the ternary expressions and I don't think the logic is as clear.

Reply:
If it were me I'd just format these on separate lines rather than trying to come up with a ternary that works with this, but as it is right now it won't pass.

Reply:
That's what I did, but I grumbled a bit as I dit it. ;)